feat: add Python MetricsRegistry Python metrics registration

keivenchang opened the pull request:
#### Overview:

Add Python MetricsRegistry with typed Prometheus metrics, enabling Python code to create and manage metrics dynamically through the Dynamo runtime. Includes introspection methods and configurable test isolation for parallel/serial execution.

#### Details:

- **Python Metrics API**: Complete Prometheus metric types (Counter, Gauge, Histogram, and Vec variants) with automatic registration to Dynamo runtime hierarchy
- **Introspection methods**: Add name(), const_labels(), and variable_labels() methods to all metric types using Prometheus Collector trait
- **Examples**: Add server_with_callback.py (callback-based) and server_with_loop.py (thread-based) with comprehensive README
- **Test infrastructure**: Add pytest suite with ENABLE_ISOLATED_ETCD_AND_NATS env var
  - Parallel mode: random ports + unique temp directories for full isolation
  - Serial mode: default ports with strict enforcement

#### Where should the reviewer start?

- lib/bindings/python/examples/metrics/README.md - Overview and usage examples
- lib/bindings/python/rust/prometheus_metrics.rs - Core Python bindings (simplified from OnceCell)
- lib/bindings/python/tests/test_metrics_registry.py - Test coverage with introspection validation
- lib/bindings/python/tests/conftest.py - Test isolation infrastructure

#### Related Issues:

DIS-734

/coderabbit profile chill

keivenchang committed:
feat: add Python dynamic registration with RuntimeMetrics

Add typed Prometheus metrics via RuntimeMetrics with endpoint.metrics.create_*() methods.
Supports all Prometheus types with automatic registration and optional callbacks.

Examples: server_with_callback.py (callback updates) and server_with_loop.py (thread updates).
Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

rmccorm4 commented on lib/bindings/python/rust/metrics.rs:
```diff
@@ -0,0 +1,269 @@
+/// Helper function to create a DynamoPromMetric - this is the main API function
+#[pyfunction]
+#[pyo3(signature = (name, metric_type, initial_value=None, label_names=None, labels=None))]
+pub fn prom_metric(
```
What about histograms? I believe you'll need to expose bucket values and such to create histograms

keivenchang replied:
Adding that in the upcoming PR.

keivenchang committed:
refactor: improve Python metrics API organization and cleanup

Move RuntimeMetrics implementation to prometheus_metrics.rs for better code organization. Update callback registration to use endpoint.metrics.register_update_callback(). Remove debug print statements from metrics subsystem (Python bindings, distributed runtime, and system status server).

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
remove unneeded import

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
update mermaid

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
update mermaid

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on lib/bindings/python/rust/metrics.rs:
```diff
@@ -0,0 +1,269 @@
+                    Ok(())
+                } else {
+                    Err("IntGaugeVec requires labels".to_string())
+                }
```
Outdated.

keivenchang commented on lib/bindings/python/rust/metrics.rs:
```diff
@@ -0,0 +1,269 @@
+        // Set initial value if provided
+        if let Some(value) = initial_value {
+            metric.set(value, labels)?;
+        }
```
Outdated.

keivenchang committed:
test: improve test isolation with configurable NATS/ETCD setup

- Add ENABLE_ISOLATED_ETCD_AND_NATS env var to control test isolation mode
- Implement parallel mode: random ports + unique temp dirs for isolation
- Implement serial mode: default ports with strict isolation checks
- Add smart port availability checking with configurable timeouts
- Add robust subprocess teardown with timeout and fallback to kill
- Add comprehensive documentation for both execution modes

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
refactor: simplify Python metrics with introspection methods

- Remove OnceCell usage, use direct Prometheus type storage
- Add introspection methods: name(), const_labels(), variable_labels()
- Remove unused MetricBase struct and Arc<Mutex<Option<T>>> wrapper
- Add GaugeVec support in MetricsRegistry trait
- Update Python type stubs to match new method signatures
- Update README with introspection examples and better organization
- Remove debug print statements from system_status_server

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
minor comment updates

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on lib/bindings/python/rust/prometheus_metrics.rs:
```diff
@@ -0,0 +1,881 @@
+use std::sync::Arc;
+
+use crate::rs;
+
```
Everything here is just boilerplate (dumb code) to allow Python to access Rust Prometheus objects. As explained in the comment below, Prometheus objects can't simply be exposed via PyO3 for reasons mentioned below.

keivenchang committed:
more minor fixes/cosmetic

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

kthui reviewed (approved):
> @kthui I noticed that the tests were running very slowly due to the sleep(5) when serially starting/stopping etcd/natsd, so I modified the code such that each test will have its own etcd/natsd listening on different ports, for 100% test isolation. This also means that they can run in parallel.

LGTM on the etcd/nats isolation in conftest.py!

keivenchang committed:
fix: eliminate etcd port race condition using port 0

Use etcd's port 0 support to let OS assign available ports, eliminating
race conditions. NATS retains retry logic (5 attempts). Adds performance
benchmarks showing isolated parallel tests (-n 16) are 1.8x faster than
default ports and 3.8x faster than isolated sequential.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

nnshah1 commented on lib/bindings/python/tests/conftest.py:
```diff
@@ -13,31 +13,347 @@
+
+
+@pytest.fixture(scope="module", autouse=True)
+def nats_and_etcd():
```
https://github.com/ai-dynamo/dynamo/blob/d809906e62840e5a566c69bb364ececf56dba11b/tests/conftest.py#L218 

We also have this being used for other tests - can we combine?

keivenchang commented on lib/bindings/python/examples/metrics/README.md:
```diff
@@ -0,0 +1,429 @@
+)
+gpu_usage: Gauge = endpoint.metrics.create_gauge(
+    "gpu_cache_usage_percent", "GPU cache usage percentage"
+)
```
Method 1: I have a ForwardPassMetrics structure/dict, and I need to populate them before calling "publish" (which sends that entire serialized structure to Rust).

Method 2: I have to add each individual item myself, but I don't have to pass in any data structures/dict. The example gpu_usage will contain an interval value, which Python can incr/decr/set, and Rust will get that automatically without a user calling publish.

keivenchang committed:
cargo fmt

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Make sure that metrics_runtime will have unique endpoints

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: handle event loop in both sync and async contexts in get_runtime()

The get_runtime() function in test_router_e2e_with_mockers.py was failing
with 'RuntimeError: no running event loop' when called from synchronous
contexts. This occurred because it unconditionally called
asyncio.get_running_loop(), which only works in async contexts.

This fix adds proper handling for both scenarios:
- In async contexts: uses asyncio.get_running_loop()
- In sync contexts: creates a new event loop with asyncio.new_event_loop()

This allows the function to work correctly whether it's called from sync
test code or from within async functions.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
test: add conditional pytest-forked requirement for test isolation

Add two modes for running tests to prevent 'Worker already initialized' errors:

1. Isolated Mode (ENABLE_ISOLATED_ETCD_AND_NATS=1):
   - Each test module gets fresh NATS/ETCD on random ports
   - Requires pytest-forked for process isolation
   - Tests using runtime fixture must have @pytest.mark.forked
   - Enables parallel test execution with pytest-xdist

2. Default Ports Mode (ENABLE_ISOLATED_ETCD_AND_NATS=0, default):
   - All test modules share NATS/ETCD on default ports (4222, 2379)
   - No pytest-forked required, no forking needed
   - Simpler setup, faster for sequential runs

Key changes:
- Add conditional pytest-forked check (only when isolated mode enabled)
- Add runtime fixture validation to enforce @pytest.mark.forked in isolated mode
- Fix mypy errors with None checks for stderr
- Use triple-quote style for multi-line error messages
- Default to ENABLE_ISOLATED_ETCD_AND_NATS=0 for simplicity
- Keep nats_and_etcd at module scope (shared per test module)

This prevents 'Worker already initialized' errors while giving users flexibility
to choose between simplicity (default) and maximum isolation (opt-in).

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: swap pytest marker order to enable test forking

Swap @pytest.mark.forked and @pytest.mark.asyncio decorator order to fix
'Worker already initialized' errors in CI. The correct order is:
  @pytest.mark.asyncio (first)
  @pytest.mark.forked (second)

When pytest-forked is before pytest-asyncio, the asyncio plugin takes precedence
and prevents tests from actually forking, causing the DistributedRuntime singleton
to persist across tests in the same process.

Also update conftest.py to gracefully reuse existing NATS/ETCD services on default
ports instead of erroring out. Added TODO to enforce proper isolation in future.

Changes:
- Swap marker order in all 11 tests in test_metrics_registry.py
- Allow reusing existing NATS/ETCD on default ports (with TODO)
- Skip teardown when reusing existing services

This matches the working pattern in test_http_server.py and fixes all test failures.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: resolve test isolation issues in test_metrics_registry and test_kv_bindings

- Changed test_kv_bindings.py distributed_runtime fixture from module to function scope
- Added @pytest.mark.forked to all tests in test_kv_bindings.py for proper isolation
- Added runtime.shutdown() to properly clean up DistributedRuntime instances
- Modified test_metrics_registry.py to create unique endpoint names per test
- All tests now use @pytest.mark.forked to avoid DistributedRuntime singleton conflicts

Fixes 'Worker already initialized' errors when running multiple test files together.
All 21 tests now pass (1 skipped) when running test_http_server.py, test_kv_bindings.py,
and test_metrics_registry.py together.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>
